File: src/main/java/dev/todaka/robustredis/protocol/CommandInput.java
package dev.todaka.robustredis.protocol;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufUtil;
import lombok.Data;

import java.util.Collections;
import java.util.List;

@Data
public class CommandInput {
    private final CommandName command;
    private List<String> keys = null;
    private List<String> args = null;

    public CommandInput(CommandName command) {
        this.command = command;
    }

    public CommandInput addKey(String key) {
        keys = Collections.singletonList(key);
        return this;
    }

    public CommandInput addKey(List<String> keys) {
        this.keys = keys;
        return this;
    }

    public CommandInput addArg(String arg) {
        args = Collections.singletonList(arg);
        return this;
    }

    public CommandInput addArg(List<String> args) {
        this.args = args;
        return this;
    }

    public void writeToByteBuf(ByteBuf buf) {
        final var len = 1 + (keys != null ? keys.size() : 0) + (args != null ? args.size() : 0);
        ByteBufUtil.writeAscii(buf, '*' + Integer.toString(len) + "\r\n");
        writeBulkString(buf, command.toString());
        if (keys != null) {
            for (final var key : keys) {
                writeBulkString(buf, key);
            }
        }
        if (args != null) {
            for (final var arg : args) {
                writeBulkString(buf, arg);
            }
        }
    }

    private void writeBulkString(ByteBuf buf, String str) {
        final var len = ByteBufUtil.utf8Bytes(str);
        ByteBufUtil.writeAscii(buf, '$' + Integer.toString(len) + "\r\n");
        ByteBufUtil.writeUtf8(buf, str);
        ByteBufUtil.writeAscii(buf, "\r\n");
    }
}

File: src/main/kotlin/dev/todaka/robustredis/CommandInput.kt
package dev.todaka.robustredis

import io.netty.buffer.ByteBuf
import io.netty.buffer.ByteBufUtil

class CommandInput(val command: CommandName) {
    var keys: List<String>? = null
    var args: List<String>? = null

    fun addKey(key: String): CommandInput {
        keys = listOf(key)
        return this
    }

    fun addKey(keys: List<String>?): CommandInput {
        this.keys = keys
        return this
    }

    fun addArg(arg: String): CommandInput {
        args = listOf(arg)
        return this
    }

    fun addArg(args: List<String>?): CommandInput {
        this.args = args
        return this
    }

    fun writeToByteBuf(buf: ByteBuf) {
        val len = 1 + (if (keys != null) keys!!.size else 0) + if (args != null) args!!.size else 0
        ByteBufUtil.writeAscii(buf, "*$len\r\n")
        writeBulkString(buf, command.toString())
        if (keys != null) {
            for (key in keys!!) {
                writeBulkString(buf, key)
            }
        }
        if (args != null) {
            for (arg in args!!) {
                writeBulkString(buf, arg)
            }
        }
    }

    private fun writeBulkString(buf: ByteBuf, str: String) {
        val len = ByteBufUtil.utf8Bytes(str)
        ByteBufUtil.writeAscii(buf, "$$len\r\n")
        ByteBufUtil.writeUtf8(buf, str)
        ByteBufUtil.writeAscii(buf, "\r\n")
    }
}
